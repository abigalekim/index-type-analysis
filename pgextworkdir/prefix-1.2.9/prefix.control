# prefix extension
comment = 'Prefix Range module for PostgreSQL'
default_version = '1.2.0'
module_pathname = '$libdir/prefix'
relocatable = true
