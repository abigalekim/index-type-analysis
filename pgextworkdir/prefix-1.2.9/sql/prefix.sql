set client_min_messages = warning;
/* reset e_f_d to 0 in PG12+ */
set extra_float_digits = 0;

create table prefixes (
       prefix    text primary key,
       name      text not null,
       shortname text,
       state     char default 'S',

       check( state in ('S', 'R') )
);
comment on column prefixes.state is 'S: assigned - R: reserved';

\copy prefixes from 'prefixes.fr.csv' with delimiter ';' csv quote '"'

create table ranges as select prefix::prefix_range, name, shortname, state from prefixes ;
create index idx_prefix on ranges using gist(prefix gist_prefix_range_ops);
analyze ranges;

set enable_seqscan to off;
select * from ranges where prefix @> '0146640123';
select * from ranges where prefix @> '0100091234';

set enable_seqscan to on;
select * from ranges where prefix @> '0146640123';
select * from ranges where prefix @> '0100091234';

select a, b, pr_penalty(a::prefix_range, b::prefix_range)
  from (values('095[4-5]', '0[8-9]'),
              ('095[4-5]', '0[0-9]'),
              ('095[4-5]', '[0-3]'),
              ('095[4-5]', '0'),
              ('095[4-5]', '[0-9]'),
              ('095[4-5]', '0[1-5]'),
              ('095[4-5]', '32'),
              ('095[4-5]', '[1-3]')) as t(a, b)
order by 3 asc;

create table numbers(number text primary key);

insert into numbers
  select '01' || substr(regexp_replace(md5(i::text), '[a-f]', '', 'g'), 1, 8)
   from generate_series(1, 5000) i;
analyze numbers;

select count(*) from numbers n join ranges r on r.prefix @> n.number;

reset client_min_messages;

-- Debian Bug 690160 regarding the symetry of <@ and @>
SELECT count(*) FROM ranges WHERE prefix <@ '01000';
SELECT count(*) FROM ranges WHERE prefix @> '01000';
SELECT count(*) FROM ranges WHERE '01000' <@ prefix;
SELECT count(*) FROM ranges WHERE '01000' @> prefix;
SELECT count(*) FROM ranges WHERE '010009888' @> prefix;
SELECT count(*) FROM ranges WHERE '010009888' <@ prefix;
SELECT count(*) FROM ranges WHERE prefix @> '010009888';
SELECT count(*) FROM ranges WHERE prefix <@ '010009888';
