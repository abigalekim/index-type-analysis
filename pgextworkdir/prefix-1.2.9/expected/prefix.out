set client_min_messages = warning;
/* reset e_f_d to 0 in PG12+ */
set extra_float_digits = 0;
create table prefixes (
       prefix    text primary key,
       name      text not null,
       shortname text,
       state     char default 'S',
       check( state in ('S', 'R') )
);
comment on column prefixes.state is 'S: assigned - R: reserved';
\copy prefixes from 'prefixes.fr.csv' with delimiter ';' csv quote '"'
create table ranges as select prefix::prefix_range, name, shortname, state from prefixes ;
create index idx_prefix on ranges using gist(prefix gist_prefix_range_ops);
analyze ranges;
set enable_seqscan to off;
select * from ranges where prefix @> '0146640123';
 prefix |      name      | shortname | state 
--------+----------------+-----------+-------
 0146   | FRANCE TELECOM | FRTE      | S
(1 row)

select * from ranges where prefix @> '0100091234';
 prefix |    name    | shortname | state 
--------+------------+-----------+-------
 010009 | LONG PHONE | LGPH      | S
(1 row)

set enable_seqscan to on;
select * from ranges where prefix @> '0146640123';
 prefix |      name      | shortname | state 
--------+----------------+-----------+-------
 0146   | FRANCE TELECOM | FRTE      | S
(1 row)

select * from ranges where prefix @> '0100091234';
 prefix |    name    | shortname | state 
--------+------------+-----------+-------
 010009 | LONG PHONE | LGPH      | S
(1 row)

select a, b, pr_penalty(a::prefix_range, b::prefix_range)
  from (values('095[4-5]', '0[8-9]'),
              ('095[4-5]', '0[0-9]'),
              ('095[4-5]', '[0-3]'),
              ('095[4-5]', '0'),
              ('095[4-5]', '[0-9]'),
              ('095[4-5]', '0[1-5]'),
              ('095[4-5]', '32'),
              ('095[4-5]', '[1-3]')) as t(a, b)
order by 3 asc;
    a     |   b    | pr_penalty  
----------+--------+-------------
 095[4-5] | 0[8-9] | 1.52588e-05
 095[4-5] | 0[0-9] | 1.52588e-05
 095[4-5] | [0-3]  |  0.00390625
 095[4-5] | 0      |  0.00390625
 095[4-5] | [0-9]  |  0.00390625
 095[4-5] | 0[1-5] |   0.0078125
 095[4-5] | 32     |           1
 095[4-5] | [1-3]  |           1
(8 rows)

create table numbers(number text primary key);
insert into numbers
  select '01' || substr(regexp_replace(md5(i::text), '[a-f]', '', 'g'), 1, 8)
   from generate_series(1, 5000) i;
analyze numbers;
select count(*) from numbers n join ranges r on r.prefix @> n.number;
 count 
-------
  2019
(1 row)

reset client_min_messages;
-- Debian Bug 690160 regarding the symetry of <@ and @>
SELECT count(*) FROM ranges WHERE prefix <@ '01000';
 count 
-------
     9
(1 row)

SELECT count(*) FROM ranges WHERE prefix @> '01000';
 count 
-------
     0
(1 row)

SELECT count(*) FROM ranges WHERE '01000' <@ prefix;
 count 
-------
     0
(1 row)

SELECT count(*) FROM ranges WHERE '01000' @> prefix;
 count 
-------
     9
(1 row)

SELECT count(*) FROM ranges WHERE '010009888' @> prefix;
 count 
-------
     0
(1 row)

SELECT count(*) FROM ranges WHERE '010009888' <@ prefix;
 count 
-------
     1
(1 row)

SELECT count(*) FROM ranges WHERE prefix @> '010009888';
 count 
-------
     1
(1 row)

SELECT count(*) FROM ranges WHERE prefix <@ '010009888';
 count 
-------
     0
(1 row)

